url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/MvdTool.java

```diff
@@ -0,0 +1,975 @@
+				{
+					MVDFile.externalise( mvd, m, folderId, 
+						Utilities.loadDBProperties(dbConn) );
+				}
+				MVDError.log("Unique percentage="+mvd.getUniquePercentage(version));
+				MvdTool.out.println(mvd.getUniquePercentage(version));
+			}
+			else
+				throw new FileNotFoundException(
+					"Couldn't find file "+t.getAbsolutePath() );
+		}
+		catch ( Exception e )
+		{
+			e.printStackTrace();
+			throw new MVDToolException( e );
+		}
+	}
+	/**
+	 * Write out all the versions as separate files
+	 */
+	private static void doArchive() throws MVDException
+	{
+		try
+		{
+			MVD mvd = loadMVD();
+			File m = new File( mvdFile );
+			File archiveDir;
+			XMLGuideFile guideFile = new XMLGuideFile( mvd );
+			// create default archive name
+			if ( archiveName == null )
+			{
+				archiveName = m.getName();
+				int index = archiveName.lastIndexOf(".");
+				archiveName = archiveName.substring( 0, index );
+				archiveDir = new File( m.getParentFile(), archiveName );
+			}
+			else
+				archiveDir = new File( archiveName );
+			if ( !archiveDir.exists() )
+				archiveDir.mkdir();
+			int nVersions = mvd.numVersions();
+			for ( short i=1;i<=nVersions;i++ )
+			{
+				char[] data = mvd.getVersion( i );
+				// use the short name as the file name
+				String vName = mvd.getVersionShortName( i );
+				File versionFile = new File( archiveDir, vName );
+                FileOutputStream fos = new FileOutputStream( versionFile );
+				OutputStreamWriter osw = new OutputStreamWriter( fos, mvd.getEncoding() );
+				osw.write( data );
+				osw.close();
+				guideFile.setVersionFile( i, vName );
+			}
+			guideFile.externalise( new File(archiveDir,XMLGuideFile.GUIDE_FILE) );
+		}
+		catch ( Exception e )
+		{
+			throw new MVDToolException( e );
+		}
+	}
+	/**
+	 * Read the versions of an archive back in to create an MVD
+	 * in one step.
+	 */
+	private static void doUnarchive() throws MVDException
+	{
+		try
+		{
+			if ( archiveName == null )
+				throw new MVDToolException("No archive folder specified!");
+			File archiveDir = new File( archiveName );
+			File guideFile = new File( archiveDir, XMLGuideFile.GUIDE_FILE );
+			if ( !guideFile.exists() )
+				throw new MVDToolException("No guide file found in "
+						+guideFile.getAbsolutePath());
+			// create empty MVD
+			if ( mvdFile == null )
+				throw new MVDToolException("No MVD file name supplied!");
+			XMLGuideFile guide = XMLGuideFile.internalise( guideFile );
+			MVD mvd = new MVD();
+            mvd.setDirectAlign( directAlignOnly );
+			mvd.setDescription( guide.getDescription() );
+			// go through the files, adding versions to the MVD
+			String[] files = guide.getVersionFileNames();
+			for ( int i=0;i<files.length;i++ )
+			{
+				short vId = (short)(i+1);
+				File versionFile = new File( archiveDir, files[i] );
+				if ( !versionFile.exists() )
+					throw new MVDToolException("File "+files[i]+" not found!");
+				FileInputStream fis = new FileInputStream( versionFile );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fis*. The resource is referenced at line 666. The resource is closed at line 667. There are other execution paths that do not contain closure statements, for example, when FileInputStream.read throws an exception. Ensure *fis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).